packages/react-import-remote/src/ImportRemote.tsx

```diff
@@ -1,111 +1,59 @@
 import * as React from 'react';
 import {Preconnect} from '@shopify/react-html';
-import {
-  RequestIdleCallbackHandle,
-  WindowWithRequestIdleCallback,
-  DeferTiming,
-} from '@shopify/async';
-import {
-  IntersectionObserver,
-  UnsupportedBehavior,
-} from '@shopify/react-intersection-observer';
-import load from './load';
+import {DeferTiming} from '@shopify/async';
+
+import {useImportRemote, Status} from './hooks';
 
 export interface Props<Imported = any> {
   source: string;
   nonce?: string;
   preconnect?: boolean;
-  onError(error: Error): void;
   getImport(window: Window): Imported;
-  onImported(imported: Imported): void;
+  onImported(imported: Imported | Error): void;
   defer?: DeferTiming;
 }
 
-interface State {
-  loaded: boolean;
-  loading: boolean;
-}
-
-export default class ImportRemote extends React.PureComponent<Props, State> {
-  state: State = {loaded: false, loading: false};
-  private idleCallbackHandle?: RequestIdleCallbackHandle;
-
-  componentWillUnmount() {
-    if (this.idleCallbackHandle != null && 'cancelIdleCallback' in window) {
-      (window as WindowWithRequestIdleCallback).cancelIdleCallback(
-        this.idleCallbackHandle,
-      );
-    }
-  }
-
-  async componentDidMount() {
-    const {defer = DeferTiming.Mount} = this.props;
-
-    if (defer === DeferTiming.Idle) {
-      if ('requestIdleCallback' in window) {
-        this.idleCallbackHandle = (window as WindowWithRequestIdleCallback).requestIdleCallback(
-          this.loadRemote,
-        );
-      } else {
-        this.loadRemote();
+export default function ImportRemote<T = unknown>({
+  source,
+  nonce,
+  preconnect,
+  getImport,
+  onImported,
+  defer,
+}: Props<T>) {
+  const {result, intersectionRef} = useImportRemote<T>(source, {
+    defer,
+    nonce,
+    getImport,
+  });
+
+  const intersectionObserver =
+    defer === DeferTiming.InViewport && intersectionRef ? (
+      <div ref={intersectionRef as React.RefObject<HTMLDivElement>} />
+    ) : null;
+
+  React.useEffect(
+    () => {
+      switch (result.status) {
+        case Status.Failed:
+          onImported(new Error(result.error.message));
```

Comment:
Why not just pass the error directly?

Reply:
It's not an instance of `Error`, it is an `Event`.

Reply:
Why is it not a real error?

Reply:
Your types seem to indicate it is a real error?